### ISSUE TITLE
[Collect approvers] [$1000]  Expense - Error appears when admin tries to pay a request without having approved it

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#34857]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.28-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
Pre-requisite: user must have created a collect workspace and
 added a BA. With an employee account, create a request and submit it
1. As the admin on ND, navigate to the workspace chat with the employee
2. Click on the expense to go to report page
3. Click on "Pay with Expensify"

## Expected Result:
It should be possible to pay the expense without problems
.
## Actual Result:
Admin can not pay the expense without approving it first, an error message appears
Unexpected error, please try again later
An unexpected error ocurred. Please try again later

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/ea8d4508-41a6-4505-a901-ede5918c721a

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0130f77ed31ac8823b</li>
        <li>Upwork Job ID: 1748765068058488832</li>
        <li>Last Price Increase: 2024-04-06</li>
    </ul>
</details>

**Comment by lanitochka17** [comment]
We think that this bug might be related to #wave7. 
CC @RachCHopkins

**Comment by promisingcoder** [comment]
Proposal
Problem Statement:
In the application, an admin cannot pay an expense without first approving it. This issue arises during the execution of the confirm function in the MoneyRequestConfirmationList component. The function terminates early if the selectedParticipants array is empty, which is often the case for admin users who are processing the payment but are not participants in the traditional sense.

Root Cause Analysis:
The root cause lies in the conditional check within the confirm function:

javascript
Copy code
if (_.isEmpty(selectedParticipants)) {
    return;
}
This condition is designed to prevent the progression of the payment process if no participants are selected. However, in the context of an admin user, this check is too restrictive, as the admin is not a participant but still needs to initiate the payment.

Proposed Solution:
The solution involves modifying the conditional check to accommodate admin users. This can be done by introducing an additional state or prop to indicate whether the current user is an admin. Based on this, the function can bypass the participant check for admins.

Updated Code:

Introduce an Admin Flag: Add a new state or prop to the component to indicate whether the user is an admin. For simplicity, let's assume a prop isAdmin is passed to the component.

Modify the Conditional Check:
Update the condition in the confirm function to bypass the participant check if the user is an admin.

javascript
Copy code
const confirm = useCallback(
    (paymentMethod) => {
        if (!props.isAdmin && _.isEmpty(selectedParticipants)) {
            return;
        }

        // ...rest of the function
    },
    [
        // ...dependencies
    ],
);
Rationale:
This modification ensures that the payment process is not unnecessarily halted for admin users. By checking the isAdmin flag, the function can appropriately handle scenarios where an admin needs to make a payment without being a participant. This change is targeted and should not cause regressions in other parts of the application, as it specifically addresses the admin scenario. Additionally, it aligns with the general expectation that admins have different permissions and capabilities compared to regular users.

Testing Considerations:
Post-implementation, thorough testing is required to ensure the payment process works correctly for both admin and non-admin users. This includes testing various scenarios and ensuring that the application's behavior remains consistent and as expected.

**Comment by promisingcoder** [comment]
Problem Statement
In the application, there is an issue where an admin cannot pay an expense without first having approved it. This occurs during the execution of the confirm function in the MoneyRequestConfirmationList component, which prematurely exits when the selectedParticipants array is empty - a common scenario for admin users.

Root Cause Analysis
The core issue is traced to the conditional check within the confirm function:

if (_.isEmpty(selectedParticipants)) {
    return;
}
This logic is intended to prevent the payment process from continuing without selected participants. However, it does not account for scenarios where an admin, who is not a participant in the traditional sense, needs to initiate the payment.

Proposed Solution
We propose modifying the conditional check to allow admin users to proceed with the payment process even when the selectedParticipants array is empty. This can be achieved by introducing an additional flag to indicate the admin status of the user.

Updated Code:

Admin Flag Introduction: Introduce a new prop, isAdmin, into the component to indicate the admin status of the user.

Conditional Check Modification:
Update the confirm function to bypass the participant check if the user is an admin:


const confirm = useCallback(
    (paymentMethod) => {
        if (!props.isAdmin && _.isEmpty(selectedParticipants)) {
            return;
        }

        // ...rest of the function
    },
    [
        // ...dependencies
    ],
);
Rationale
This adjustment ensures that the payment process for admin users is not obstructed by the participant check. The isAdmin flag allows the function to handle cases where an admin is required to make a payment without being a listed participant. This targeted modification should not cause any regressions elsewhere in the application, as it specifically caters to the unique role of admin users.

Testing and Validation
After implementing this change, it is crucial to conduct comprehensive testing to confirm that the payment process functions correctly for both admin and non-admin users in various scenarios. This will help ensure the application behaves as expected and maintains consistency across different user roles.

**Comment by abdulrahuman5196** [comment]
Will check this issue in my morning.

**Comment by mallenexpensify** [comment]
@abdulrahuman5196 please review the above. @RachCHopkins I agree these seems to fit with Wave 7 so I've added to the project there. Let me know if you disagree.

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
@promisingcoder Could you kindly update the proposal as per the proposal template - https://raw.githubusercontent.com/Expensify/App/main/contributingGuides/PROPOSAL_TEMPLATE.md

Meanwhile I did a high level check. Could you kindly provide more information on how `MoneyRequestConfirmationList` is related to the action taken on the SettlementButton ?

**Comment by abdulrahuman5196** [comment]
@mallenexpensify Could you kindly provide information on how to create this pre-requisite?

```
Pre-requisite: user must have created a collect workspace and
added a BA. With an employee account, create a request and submit it
```

**Comment by promisingcoder** [comment]
@abdulrahuman5196 
MoneyRequestConfirmationList: This is a list that shows all the participants involved in a money request. It also handles the confirmation of the request.

SettlementButton: This is a button that, when pressed, triggers the payment process. It's used within the MoneyRequestConfirmationList component.

The action of pressing the SettlementButton triggers a function called confirm in the MoneyRequestConfirmationList component. This function checks if there are any selected participants and if a payment method is provided. If these conditions are met, it sets a state called didConfirm to true and calls another function, onSendMoney, with the selected payment method as an argument.

The proposal suggests modifying the confirm function to accommodate admin users who need to confirm payments but are not participants in the traditional sense. Currently, the confirm function terminates early if the selectedParticipants array is empty, which is often the case for admin users. By introducing an additional check for admin users in the confirm function, the SettlementButton can successfully trigger the payment process even when the user is an admin.

So, in simple terms, the SettlementButton starts the action that the confirm function in the MoneyRequestConfirmationList component carries out. The proposed changes aim to make sure that this process works correctly for all users, including admins.

**Comment by mallenexpensify** [comment]
> Pre-requisite: user must have created a collect workspace and
added a BA. With an employee account, create a request and submit it

@abdulrahuman5196 , most of that can be done view OldDot, expensify.com.  The part I'm unsure about is 'added a BA` (which I'm assuming is Bank Account) Ah ha, looks like this should work 
https://expensify.slack.com/archives/C01GTK53T8Q/p1706551410056249?thread_ts=1706541901.752409&cid=C01GTK53T8Q

**Comment by abdulrahuman5196** [comment]
Will check on the old dot setup in my morning

**Comment by abdulrahuman5196** [comment]
Still having trouble with setup, will try today or reach out in slack

**Comment by abdulrahuman5196** [comment]
@mallenexpensify These are the steps I followed
1) I created a new account abdulrahuman5196+collect@gmail.com
2) Created a  old dot collect workspace  (abdulrahuman5196+OldDotcollect)
3) Created a new workspace in new dot (Test new dot)
4) Add bank account as per the steps in slack suggested.
5) Added employee for the old dot workspace and created a report (abdulrahuman5196+4@gmail.com)

But in the new dot, I don't see a workspace chat for the **abdulrahuman5196+OldDotcollect** itself. Not sure if I am missing something here. Tried different ways, but for the collect workspaces I create in old dot, I don't get workspace chat in new dot, only `announce` and `admin` is present for the workspaces created in old dot. 

Attached the screenshots for reference. Do let me know if I am doing some configuration wrongly.



<img width="1512" alt="Screenshot 2024-02-09 at 12 03 04 AM" src="https://github.com/Expensify/App/assets/46707890/28b472a3-206c-48af-ad8b-e7ca8c3d58f2">
<img width="1511" alt="Screenshot 2024-02-09 at 12 03 20 AM" src="https://github.com/Expensify/App/assets/46707890/36cf6bd0-3890-442b-ab31-5a9c60e93d88">
<img width="1510" alt="Screenshot 2024-02-09 at 12 03 33 AM" src="https://github.com/Expensify/App/assets/46707890/ae317659-a96f-4e67-8bf1-abc79def6430">
<img width="1512" alt="Screenshot 2024-02-09 at 12 03 40 AM" src="https://github.com/Expensify/App/assets/46707890/d4773ce6-377d-4305-9ca6-59d4b6d19a12">
<img width="1512" alt="Screenshot 2024-02-09 at 12 04 52 AM" src="https://github.com/Expensify/App/assets/46707890/61c9f7a8-11ec-4eda-ba1c-af36eee0dba8">

**Comment by mallenexpensify** [comment]
Thanks for all the detail and steps @abdulrahuman5196 , I'm checking in the Wave 7 room [here](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1707509779241739?thread_ts=1707452633.267759&cid=C02NZ2HGJAZ) to see if you need to be added to any betas.

**Comment by mallenexpensify** [comment]
@abdulrahuman5196 I just added you to a collect plan to test with, `Collect policy Feb 12 2024`, where abdulrahuman5196+4@gmail.com should be an employee and abdulrahuman5196+collect@gmail.com should be an admin. 

Can you take a look? It's the first time doing this for me so I'm unsure everything's right.

**Comment by marcochavezf** [comment]
Something interesting in the video is that we don’t show the Pay button after the report is approved, and we do that when auto-reimbursement is enabled in the policy. If that’s the case, the problem is that we’re showing the Pay button when we shouldn’t.

**Comment by mallenexpensify** [comment]
Bumping to $1k, with the hope of getting more proposals.

**Comment by promisingcoder** [comment]
Proposal

Problem:

Expensify users who attempt to make IOU payments are sometimes encountering a generic 'iou.error.other' message. This indicates a failure in the payment process, but the lack of a specific error makes it difficult to diagnose and frustrating for the user.

Root Cause:

While the exact root cause is still under investigation, the strong  suspicion is a mismatch between optimistic update assumptions and real-time payment capabilities.  Here's why we think this:

KYCWall Integration: The payment flow heavily interacts with the KYCWall component, which enforces Know Your Customer checks and verifies payment methods.
Generic Error Handling: It's likely that specific errors (payment method setup incomplete, KYC failed, etc.) are not propagating correctly when optimistic updates are created. Instead, they fall back to 'iou.error.other'.
Changes to Solve the Problem

Granular Error Handling:

Modify the KYCWall component to generate descriptive error objects in scenarios where payment cannot proceed (e.g., no verified payment method, failed KYC).
Ensure these detailed errors are passed back to functions like getSendMoneyParams so they populate failureData keys instead of 'iou.error.other'.
Conditional Optimistic Updates:

Refactor getSendMoneyParams to conditionally create optimistic updates ONLY after user has passed required payment and KYC checks in KYCWall.

Introduce potential use of an Onyx key that flags whether KYC/payment setup is complete, triggering the generation of optimistic updates when true.

Enhanced Logging:

In a development/staging environment, add more logging around calls to getSendMoneyParams and within the KYCWall component. Focus on logging arguments, responses, and state changes at points where failures occur.
Coordinate these logs with backend logs (particularly API endpoints related to payment handling) to identify discrepancies in expected vs. actual frontend state.
Alternative Solutions Explored

Backend-Driven Fix: It's possible the root cause is misaligned backend API expectations on payment readiness that our frontend does not anticipate. However, we want to investigate better frontend error handling, as it provides valuable immediate feedback to the user regardless of server changes.
will be updated soon

**Comment by promisingcoder** [comment]
Understanding the Problem

Root Cause: The current implementation assumes that all participants involved in a request are required to approve before payment. This assumption breaks down in admin scenarios, where an admin is not a designated participant but needs to initiate the payment process.
Symptom: Admins attempting to pay a request before approval are met with an error message preventing them from completing the action.
Proposed Solution

The core modification aims to introduce an admin flag and conditional check within the confirm function of the MoneyRequestConfirmationList component, allowing payment processing by admins even in the absence of participant approval. Here's the breakdown:

Introduce an Admin Flag:

Add a new boolean state variable or pass a prop called isAdmin to the MoneyRequestConfirmationList component. This flag signals whether the current user is an admin.
Modify the Conditional Check:

Adjust the existing if (_.isEmpty(selectedParticipants)) { return; } condition within the confirm function.
Introduce a logical OR to accommodate admins using the flag: if (_.isEmpty(selectedParticipants) && !props.isAdmin) { return; } This revised condition allows the payment process to proceed if either selectedParticipants is not empty or the user is an admin.
Code Example

Assuming the admin flag is passed as a prop:

JavaScript
const MoneyRequestConfirmationList = (props) => {
    // ... other code

    const confirm = useCallback(
        (paymentMethod) => {
            if (!props.isAdmin && _.isEmpty(selectedParticipants)) {
                return;
            }
            // ... rest of the function, payment processing logic
        },
        [/* relevant dependencies */]
    );

    // ... rest of the component
};

**Comment by abdulrahuman5196** [comment]
Rechecking now.

**Comment by abdulrahuman5196** [comment]
> @abdulrahuman5196 I just added you to a collect plan to test with, `Collect policy Feb 12 2024`, where [abdulrahuman5196+4@gmail.com](mailto:abdulrahuman5196+4@gmail.com) should be an employee and [abdulrahuman5196+collect@gmail.com](mailto:abdulrahuman5196+collect@gmail.com) should be an admin.
> 
> Can you take a look? It's the first time doing this for me so I'm unsure everything's right. (also... reminder to self, if it does work, ask for any contributor's emails )

Hi @mallenexpensify , Thank you setting up this. I currently get a workspace chat for the old dot policy.

I requested couple of expenses from the employee account.

In the admin account, while I am trying to "Pay with Expensify" option, I am prompted with bank setup, which i am doing as per mentioned here https://expensify.slack.com/archives/C01GTK53T8Q/p1706551410056249?thread_ts=1706541901.752409&cid=C01GTK53T8Q. My bank account is added, but it still asks me for verification and not allows me to pay with expensify. Attached snapshots below.

Anyways I tried with "Pay with Elsewhere" option, this seems to be working fine. Can you try again with the reports I have submitted from employee account? Or any other ways to unblock on bank setup step?

<img width="1512" alt="Screenshot 2024-02-16 at 3 17 17 PM" src="https://github.com/Expensify/App/assets/46707890/02535dba-9511-41f9-8c8b-a59d3ec5654d">
<img width="1512" alt="Screenshot 2024-02-16 at 3 13 36 PM" src="https://github.com/Expensify/App/assets/46707890/143cf68e-cd33-453f-8929-1639d4ac83df">
<img width="1510" alt="Screenshot 2024-02-16 at 3 13 25 PM" src="https://github.com/Expensify/App/assets/46707890/64a1bec8-8d5d-4775-a2ae-2b068c99e135">

**Comment by kmbcook** [comment]
@mallenexpensify could you also add me to a test collect workspace, to work on this issue?

**Comment by mallenexpensify** [comment]
> Can you try again with the reports I have submitted from employee account? Or any other ways to unblock on bank setup step?

I approved the report I see, does that work/help? 
<img width="1039" alt="image" src="https://github.com/Expensify/App/assets/22508304/73c5e616-2c07-4554-8709-727a06974e8a">

> I more questions. given that "Pay with Elsewhere" seems to mark the request as paid.

I _think_ this might be a known bug that's currently being worked on. 

>Should our expectation be to Disable "Pay with" options until the "approve" option is made?

Asking in an internal room [here](https://expensify.slack.com/archives/C02NZ2HGJAZ/p1708289637528309?thread_ts=1707949088.467969&cid=C02NZ2HGJAZ)

@kmbcook added kevincook13@gmail.com to high traffic account (twas the email listed in your GH profile).

**Comment by trjExpensify** [comment]
Do we need this issue, @marcochavezf? I feel like we're going to duplicate/conflict the work with https://github.com/Expensify/App/issues/34951.

**Comment by mallenexpensify** [comment]
Put on hold pending 
- https://github.com/Expensify/App/issues/34951 

Thanks @marcochavezf and @trjExpensify !

**Comment by mallenexpensify** [comment]
Holding, [PR we're holding on ](https://github.com/Expensify/App/pull/36814)looks like it's closed to being me merged.

**Comment by mallenexpensify** [comment]
Still on hold
- https://github.com/Expensify/App/issues/34951

**Comment by trjExpensify** [comment]
Why's this one held still? The linked issue is ready for payment tomorrow?

**Comment by mallenexpensify** [comment]
Unsure @trjExpensify .  My process for held issues is to hover over the link to see if it's open or closed.  Musta misread it for this one.  

Taking off hold, Added `Retest-weekly`

**Comment by mallenexpensify** [comment]
Waiting on the retest.

**Comment by mallenexpensify** [comment]
Waiting on retest

**Comment by abdulrahuman5196** [comment]
waiting on retest

**Comment by mvtglobally** [comment]
Issue is reproducible during KI retests.


https://github.com/Expensify/App/assets/43995119/c81b7903-f482-483a-9b9b-afc222e7bfff

**Comment by mallenexpensify** [comment]
Just tried to reproduce and can't.  @abdulrahuman5196 , can you please try? 
https://github.com/Expensify/App/assets/22508304/fd6888f4-e2cc-4f24-b3f5-8cb3e3e1f8a4

**Comment by abdulrahuman5196** [comment]
> Issue is reproducible during KI retests.
> 1712169548259.cc.mp4

Actually, Don't understand this reproduced video.
I don't see any errors here? @mvtglobally

**Comment by mallenexpensify** [comment]
Unable to reproduce again. This issue is a few months old and we've had many changes. If we encounter it again, someone can create a new issue with repro steps. 

<img width="913" alt="image" src="https://github.com/Expensify/App/assets/22508304/6c9d24e4-e2b8-4f40-bcc1-38d4b9e4717e">
